{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scipy.stats\n",
    "import matplotlib.pylab as pylab\n",
    "import ipywidgets as widgets\n",
    "from os import path\n",
    "from sklearn.decomposition import PCA\n",
    "from matplotlib.offsetbox import AnchoredText\n",
    "from IPython.display import clear_output\n",
    "from IPython.display import display\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>PDX_14P3_20</th>\n",
       "      <th>PDX_14P3_20_Sim</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>5066.279722</td>\n",
       "      <td>5159.717472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>14.092113</td>\n",
       "      <td>14.901953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>1715.516365</td>\n",
       "      <td>1831.702142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>192.058293</td>\n",
       "      <td>213.403958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>365.414831</td>\n",
       "      <td>337.999417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20947</th>\n",
       "      <td>ENSG00000273485</td>\n",
       "      <td>132.876114</td>\n",
       "      <td>131.129663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20948</th>\n",
       "      <td>ENSG00000273486</td>\n",
       "      <td>118.774919</td>\n",
       "      <td>117.230137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20949</th>\n",
       "      <td>ENSG00000273487</td>\n",
       "      <td>16.105308</td>\n",
       "      <td>15.895381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20950</th>\n",
       "      <td>ENSG00000273488</td>\n",
       "      <td>45.297330</td>\n",
       "      <td>44.704621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20951</th>\n",
       "      <td>ENSG00000273492</td>\n",
       "      <td>9.223192</td>\n",
       "      <td>3.903204</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20952 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                GeneID  PDX_14P3_20  PDX_14P3_20_Sim\n",
       "0      ENSG00000000003  5066.279722      5159.717472\n",
       "1      ENSG00000000005    14.092113        14.901953\n",
       "2      ENSG00000000419  1715.516365      1831.702142\n",
       "3      ENSG00000000457   192.058293       213.403958\n",
       "4      ENSG00000000460   365.414831       337.999417\n",
       "...                ...          ...              ...\n",
       "20947  ENSG00000273485   132.876114       131.129663\n",
       "20948  ENSG00000273486   118.774919       117.230137\n",
       "20949  ENSG00000273487    16.105308        15.895381\n",
       "20950  ENSG00000273488    45.297330        44.704621\n",
       "20951  ENSG00000273492     9.223192         3.903204\n",
       "\n",
       "[20952 rows x 3 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class StatisticalPlotGenerator:\n",
    "    def __init__(self):\n",
    "        #self.__pathList_ = []\n",
    "        #self.__dataDict = {}\n",
    "        #self.__cutOption = \" \"\n",
    "        \n",
    "\n",
    "        # a dictionary, key is the sample name, value is a list which contains all the cnt matrices for the same\n",
    "        # sample\n",
    "        self.__sampleDicts = {}\n",
    "        #self.__sampleNames = \" \"\n",
    "        self.__tempDataFrame = None\n",
    "        \n",
    "        \n",
    "    def __loadRSEMReal(self, sampleName):\n",
    "        \"\"\"\n",
    "        as the name, this function aims loading data into the program. \n",
    "        files must be stored in csv format, and each csv file contains one pair or more than one pairs of the counts of\n",
    "        the samples. that means, a csv format must contain at least one of the real data's the counts and the counts of\n",
    "        a simulated data which has been generated based on the read data.\n",
    "    \n",
    "        input:\n",
    "        @sampleName: the name of the sample\n",
    "    \n",
    "        \"\"\"\n",
    "        cutOff = [0, 2.5, 5, 7.5, 10]\n",
    "        # key of the tempDict is the cut off threshould, value is the cnt matrix accordingly\n",
    "        tempDict = {}\n",
    "        dictList = []\n",
    "        for cut in cutOff:            \n",
    "            fullPath = \"/home/lima/Projects/simulation/SingleSample/Samples/{sampleName}WithReadsShorterThan{cut}cutOff.csv\".format(sampleName = sampleName, cut = cut)\n",
    "            dataFrame = pd.read_csv(fullPath)\n",
    "            dataFrame = dataFrame.rename(columns = {\"Unnamed: 0\": \"GeneID\"})\n",
    "            #print(dataFrame.isnull().values.any())\n",
    "            \n",
    "            tempDict[cut] = dataFrame\n",
    "            dictList.append(tempDict)\n",
    "        self.__sampleDicts[sampleName] = dictList\n",
    "        \n",
    "   \n",
    "    \n",
    "    def prepRSEMReal(self):\n",
    "        nameOptions = pd.read_csv(\"/home/lima/Projects/simulation/SingleSample/Option.csv\", sep=\" \", header = None)\n",
    "        for index in range(len(nameOptions)):\n",
    "            name = nameOptions.iloc[index, 0]\n",
    "            print(name)\n",
    "            self.__loadData(name)\n",
    "        #print(len(self.__sampleDicts))\n",
    "            \n",
    "    def plot(self, samples):\n",
    "        \"\"\"\n",
    "        if all the samples are plotted out, this notebook will not be able to uploade to github\n",
    "        so, just for example, i will take two samples\n",
    "        \"\"\"\n",
    "        samples = samples\n",
    "        for sample in samples:\n",
    "            cntMatrices = self.__sampleDicts[sample]\n",
    "            sim_name = \"{sampleName}_Sim\".format(sampleName = sample)\n",
    "            sampleNames = [sample, sim_name]\n",
    "            index = 0\n",
    "            for matrix in cntMatrices:\n",
    "                keys = matrix.keys()\n",
    "                values = matrix.values()\n",
    "                keys = list(keys)\n",
    "                values = list(values)\n",
    "                \n",
    "                # key is the cut off threshould\n",
    "                key = keys[index]\n",
    "                value = values[index]\n",
    "                index = index + 1\n",
    "                #print(sampleNames)\n",
    "                #print(\"Cut off threshould is {key} .....\".format(key = key))\n",
    "                #print(value.head())\n",
    "                #print(\"{number} genes left\".format(number = len(value)))\n",
    "                \n",
    "                df = value[sampleNames]\n",
    "                dfOne = np.log2(df + 1)\n",
    "                dfTwo = np.log2(df + 1)\n",
    "                pen = PlotPen()\n",
    "                pen.plotPlotBlock(dfOne, key, sampleNames)\n",
    "                pen.plotCorrScatterPlot(dfTwo, key, sampleNames)\n",
    "                \n",
    "class PlotPen:\n",
    "    \n",
    "    def plotPlotBlock(self, df, cutOff, names):\n",
    "        \n",
    "        params = {'legend.fontsize': 8, \n",
    "                    'font.family': 'serif',\n",
    "                    'font.weight': 'medium',\n",
    "                    'font.variant': 'normal',\n",
    "                    'figure.figsize': (10, 10), \n",
    "                    'figure.edgecolor': '#04253a', \n",
    "                    'figure.titlesize': 8, \n",
    "                    'axes.labelsize': 6, \n",
    "                    'axes.titlesize': 6, \n",
    "                    'xtick.labelsize': 6, \n",
    "                    'ytick.labelsize': 6}\n",
    "        pylab.rcParams.update(params)\n",
    "        figureOne, axes = plt.subplots(2, 2, constrained_layout = True)\n",
    "        self.buildPlotBlock(df, axes, cutOff, names)\n",
    "        plt.suptitle('Statistical plots for comparing {realData} & {simulatedData} with reads shorter than {cut} being removed'.format(realData = names[0], simulatedData = names[1], cut = cutOff))\n",
    "        title = 'Real data and synthetic data of {realData} with reads shorter than {cut} being removed.png'.format(realData = names[0], cut = cutOff)\n",
    "        plt.savefig(title, dpi = 300, bbox_inches = 'tight')\n",
    "        plt.plot()\n",
    "        \n",
    "    def buildPlotBlock(self, df, axes, cutOff, names):\n",
    "        \"\"\"\n",
    "        buildPlotBlock is a helper function helping putting the desired plots, such as the scatter plot, box plot,\n",
    "        violin plot and the distribution plot into a grided pyplot\n",
    "    \n",
    "        Input:\n",
    "        @df: the dataframe which provides the data\n",
    "        @axes: a 2x2 subplot matrix \n",
    "        \"\"\"\n",
    "        # df = np.log2(df[names] + 1)\n",
    "        sns.set(style = \"whitegrid\", palette = \"muted\", color_codes = True)\n",
    "        # put the violin plot to axes[0, 0]\n",
    "        axes[0, 0].clear()\n",
    "        sns.violinplot(data = df, linewidth = 0.2, palette = \"husl\", ax = axes[0, 0])\n",
    "    \n",
    "        # put the boxplot to axes[1, 1]\n",
    "        axes[1, 1].clear()\n",
    "        sns.boxplot(data = df, linewidth = 0.2, palette = \"husl\", ax = axes[1, 1])\n",
    "    \n",
    "        # put the distribution plot to axes[0, 1]\n",
    "        axes[0, 1].clear()\n",
    "        sns.distplot(df.iloc[ : , 0], kde = True, color = \"b\", label = 'Real', ax = axes[0, 1])\n",
    "        sns.distplot(df.iloc[ : , 1], kde = True, color = \"r\", label = 'Simulated', ax = axes[0, 1])\n",
    "        axes[0, 1].legend(loc = 'best', prop = {'size': 8})\n",
    "        axes[0, 1].set_xlabel(\"Distribution plot for real data and simulated data\")\n",
    "    \n",
    "        # put the scatter plot to axes [1, 0]\n",
    "        axes[1, 0].clear()\n",
    "        self.scatterPlot(df, axes[1, 0], cutOff, names)\n",
    "                         \n",
    "    def scatterPlot(self, df, ax, cutOff, names):\n",
    "        \"\"\"\n",
    "        the scatterPlot function plots a pairwise scatter plot for the simulated data and the real data for the same \n",
    "        sample, with the x-coordinates being the counts number of the real data and the y-coordinates being the counts\n",
    "        number of the simulated data\n",
    "    \n",
    "        input data:\n",
    "        @df: a pandas dataframe contains the counts for the real and the simulated data\n",
    "        @ax: an axes object\n",
    "        @names: a list contains the sample names\n",
    "        \"\"\"\n",
    "    \n",
    "        # notInrealData, notInSimulatedData, inNeither are lists stores the row indecies for data as theirs names\n",
    "        notInRealData = df.index[df.iloc[ : , 0] == 0].tolist()\n",
    "        notInSimulatedData = df.index[df.iloc[ : , 1] == 0].tolist()\n",
    "        inNeither = list(set(notInRealData) & set(notInSimulatedData))\n",
    "    \n",
    "        # inBoth, simOnly, realOnly, neither are integers for \n",
    "        # genes found in both samples, one of the samples or none of them\n",
    "        neither = len(inNeither)\n",
    "        simOnly = len(notInRealData) - neither\n",
    "        realOnly = len(notInSimulatedData) - neither\n",
    "        inBoth = len(df) - simOnly - realOnly + neither\n",
    "    \n",
    "        # category the genes into 4 classes\n",
    "        df[\"ExpressionStatus\"] = [\"BothSamples\"] * len(df)\n",
    "        df.loc[notInRealData, \"ExpressionStatus\"] = \"SimulatedOnly\"\n",
    "        df.loc[notInSimulatedData, \"ExpressionStatus\"] = \"RealOnly\"\n",
    "        df.loc[inNeither, \"ExpressionStatus\"] = \"NeithSample\"\n",
    "        # print(df)\n",
    "        # set up the plotting\n",
    "        sns.set(style = \"whitegrid\", palette = \"muted\", color_codes = True)\n",
    "        # sns.set(style = \"dark\", palette = \"inferno\", color_codes = True)\n",
    "        # calculate the correlation between the two sets of data.\n",
    "        # print(df.iloc[ : , 0])\n",
    "        col_one = df.iloc[ : , 0]\n",
    "        col_two = df.iloc[ : , 1]\n",
    "        corr = scipy.stats.pearsonr(col_one, col_two)[0]\n",
    "        \n",
    "        sns.scatterplot(x = names[0], y = names[1], hue = 'ExpressionStatus', marker = '+', style = 'ExpressionStatus', data = df, ax = ax, palette = 'inferno', s = 2.0)\n",
    "        at = AnchoredText(\"{inBoth} are found in both real and simulated data\\n{neither} expressed in neither real nor simulated data\\n{realOnly} are found in real data only\\n{simOnly} are found in simulated data only\\nCut off threshould: {cut}\\nCorr: {corr}\".format(\n",
    "            neither = neither, realOnly = realOnly, simOnly = simOnly, inBoth = inBoth, corr = corr, total = len(df) - realOnly - simOnly + len(inNeither), cut = cutOff), \n",
    "                          prop = dict(size = 8), frameon = True, loc = 'upper left')\n",
    "        at.patch.set_boxstyle(\"round,pad=0.,rounding_size=0.2\")\n",
    "   \n",
    "        ax.add_artist(at)\n",
    "        ax.set_xlabel(\"Real Data\")\n",
    "        ax.set_ylabel(\"Simulated Data\")\n",
    "        ax.legend(loc = 'lower right', fancybox = True, shadow = False, scatterpoints = 200, prop = {'size': 8})\n",
    "\n",
    "    def plotCorrScatterPlot(self, df, cutOff, names):\n",
    "        print(\"Called plotCorrScatterPlot\")\n",
    "        params = {'legend.fontsize': 8, \n",
    "                  'figure.figsize': (10, 10), \n",
    "                  'figure.titlesize': 8,\n",
    "                  'axes.labelsize': 6, \n",
    "                  'axes.titlesize': 6, \n",
    "                  'xtick.labelsize': 6, \n",
    "                  'ytick.labelsize': 6}\n",
    "        pylab.rcParams.update(params)\n",
    "        figureTwo, axes = plt.subplots()\n",
    "        self.corrScatterPlot(df, axes, cutOff, names)\n",
    "        title = 'Scattered correlation plot for the real data and synthetic data of {sample} with {cut} cut off.png'.format(sample = names[0], cut = cutOff)\n",
    "        plt.savefig(title, dpi = 300, bbox_inches = 'tight')\n",
    "        plt.plot()\n",
    "                         \n",
    "    # function corrScatterPlot is a helper function for scatter-plotting the correlation between two sets of data with \n",
    "    # different colors. \n",
    "    def corrScatterPlot(self, df, ax, cutOff, names):\n",
    "        \"\"\"\n",
    "        the corrScatterPlot is a scatter plot on which the counts of the real data and the simulated data are plotted.\n",
    "        before plotting the scatter plot, first process the data set with a 2-components PCA. picking the principal \n",
    "        component keeps the most variability and using it as the x-coordinates. then, plot the scatter plot with the\n",
    "        counts of the real data and the simulated data as the y-coordinates\n",
    "        Input:\n",
    "        @df: a pandas dataframe object which provides the data for plotting\n",
    "        @ax: an axes object on which the plot is going to be plotted\n",
    "        @names: names of samples to be plotted\n",
    "        Output:\n",
    "        an axes object\n",
    "        \"\"\"\n",
    "        pca = PCA(n_components = 2)\n",
    "        pca.fit(df)\n",
    "    \n",
    "        # take out the explained_variance_ for finding the one keeps the most variability\n",
    "        variances = pca.explained_variance_\n",
    "    \n",
    "        # index records the index of the desired variances\n",
    "        index = 0 if (variances[0] > variances[1]) else 1\n",
    "    \n",
    "        pcs = pca.transform(df)\n",
    "    \n",
    "        # take the components keep the most variability\n",
    "        pc = pcs[ : , index]\n",
    "    \n",
    "        # calculate how many genes lost in the principal component\n",
    "        diff = len(df) - len(pc)\n",
    "    \n",
    "        # find the percentage of the variability being kept in the principal component\n",
    "        var_ratio = pca.explained_variance_ratio_\n",
    "        var_most = var_ratio[index]\n",
    "    \n",
    "        # find the minimum components\n",
    "        min_pc = np.amin(pc)\n",
    "\n",
    "        # check if it is negative, make it non-negative and add it to pc\n",
    "        if min_pc < 0:\n",
    "            min_pc = np.abs(min_pc)\n",
    "            pc = pc + min_pc\n",
    "        \n",
    "        #calculate the correlation between the two sets of data.\n",
    "        corr = scipy.stats.pearsonr(df.iloc[ : , 0], df.iloc[ : , 1])[0]   \n",
    "        plt.style.use('seaborn-whitegrid')\n",
    "        # plt.style.use('dark_background')\n",
    "        ax.clear()\n",
    "        ax.scatter(pc, df.iloc[ : , 0], s = 1, c = 'b', alpha = 0.5, marker = \"s\", label = 'Real')\n",
    "        ax.scatter(pc, df.iloc[ : , 1], s = 1, c = 'r', alpha = 0.5, marker = \"o\", label = 'Simulated')\n",
    "    \n",
    "        at = AnchoredText(\"Sample: {sample}\\nCut threshould: {cut}\\nRatio of variability being kept in the principal component: {most}\\n{diff} genes lost in the principal component\\nCorr: {corr}\".format(\n",
    "            sample = names[0], corr = corr, cut = cutOff, most = var_most, diff = diff), \n",
    "                          prop = dict(size = 8), frameon = True, loc = 'upper left')\n",
    "        at.patch.set_boxstyle(\"round,pad=0.,rounding_size=0.2\")\n",
    "        ax.add_artist(at)\n",
    "        ax.set_xlabel(\"Simulated Data\")\n",
    "        ax.set_ylabel(\"Real Data\")\n",
    "        #ax.set_title(\"Scatted plot for the correlation between real data and simulated data\")\n",
    "        ax.legend(loc = 'center left', fancybox = True, shadow = False, scatterpoints = 200, prop = {'size': 8});\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TranscriptomeID</th>\n",
       "      <th>PDX_10Veh_Kall</th>\n",
       "      <th>PDX_12Veh_Kall</th>\n",
       "      <th>PDX_13P3_40_Kall</th>\n",
       "      <th>PDX_14P3_20_Kall</th>\n",
       "      <th>PDX_1P3_20_Kall</th>\n",
       "      <th>PDX_2P3_40_Kall</th>\n",
       "      <th>PDX_3P3_40_Kall</th>\n",
       "      <th>PDX_4P3_20_Kall</th>\n",
       "      <th>PDX_5P3_40_Kall</th>\n",
       "      <th>PDX_6P3_20_Kall</th>\n",
       "      <th>PDX_7P3_20_Kall</th>\n",
       "      <th>PDX_8Veh_Kall</th>\n",
       "      <th>PrimaryTumor_Kall</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000415118.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000434970.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000448914.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000604642.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000603326.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180248</th>\n",
       "      <td>ENST00000450690.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180249</th>\n",
       "      <td>ENST00000605523.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.814364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180250</th>\n",
       "      <td>ENST00000605654.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.398913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180251</th>\n",
       "      <td>ENST00000603923.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180252</th>\n",
       "      <td>ENST00000600861.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>180253 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          TranscriptomeID  PDX_10Veh_Kall  PDX_12Veh_Kall  PDX_13P3_40_Kall  \\\n",
       "0       ENST00000415118.1             0.0             0.0               0.0   \n",
       "1       ENST00000434970.2             0.0             0.0               0.0   \n",
       "2       ENST00000448914.1             0.0             0.0               0.0   \n",
       "3       ENST00000604642.1             0.0             0.0               0.0   \n",
       "4       ENST00000603326.1             0.0             0.0               0.0   \n",
       "...                   ...             ...             ...               ...   \n",
       "180248  ENST00000450690.1             0.0             0.0               0.0   \n",
       "180249  ENST00000605523.1             0.0             0.0               0.0   \n",
       "180250  ENST00000605654.1             0.0             0.0               0.0   \n",
       "180251  ENST00000603923.1             0.0             0.0               0.0   \n",
       "180252  ENST00000600861.1             0.0             0.0               0.0   \n",
       "\n",
       "        PDX_14P3_20_Kall  PDX_1P3_20_Kall  PDX_2P3_40_Kall  PDX_3P3_40_Kall  \\\n",
       "0                    0.0              0.0              0.0              0.0   \n",
       "1                    0.0              0.0              0.0              0.0   \n",
       "2                    0.0              0.0              0.0              0.0   \n",
       "3                    0.0              0.0              0.0              0.0   \n",
       "4                    0.0              0.0              0.0              0.0   \n",
       "...                  ...              ...              ...              ...   \n",
       "180248               0.0              0.0              0.0              0.0   \n",
       "180249               0.0              0.0              0.0              0.0   \n",
       "180250               0.0              0.0              0.0              0.0   \n",
       "180251               0.0              0.0              0.0              0.0   \n",
       "180252               0.0              0.0              0.0              0.0   \n",
       "\n",
       "        PDX_4P3_20_Kall  PDX_5P3_40_Kall  PDX_6P3_20_Kall  PDX_7P3_20_Kall  \\\n",
       "0                   0.0              0.0              0.0              0.0   \n",
       "1                   0.0              0.0              0.0              0.0   \n",
       "2                   0.0              0.0              0.0              0.0   \n",
       "3                   0.0              0.0              0.0              0.0   \n",
       "4                   0.0              0.0              0.0              0.0   \n",
       "...                 ...              ...              ...              ...   \n",
       "180248              0.0              0.0              0.0              0.0   \n",
       "180249              0.0              0.0              0.0              0.0   \n",
       "180250              0.0              0.0              0.0              0.0   \n",
       "180251              0.0              0.0              0.0              0.0   \n",
       "180252              0.0              0.0              0.0              0.0   \n",
       "\n",
       "        PDX_8Veh_Kall  PrimaryTumor_Kall  \n",
       "0                 0.0           0.000000  \n",
       "1                 0.0           0.000000  \n",
       "2                 0.0           0.000000  \n",
       "3                 0.0           0.000000  \n",
       "4                 0.0           0.000000  \n",
       "...               ...                ...  \n",
       "180248            0.0           0.000000  \n",
       "180249            0.0          16.814364  \n",
       "180250            0.0           1.398913  \n",
       "180251            0.0           0.000000  \n",
       "180252            0.0           0.000000  \n",
       "\n",
       "[180253 rows x 14 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kallPath = \"/home/lima/Projects/simulation/TranscriptSingleSample/cntMatrices_Kall.csv\"\n",
    "kallDF = pd.read_csv(kallPath)\n",
    "kallDF = kallDF.rename(columns = {\"Unnamed: 0\": \"TranscriptomeID\"})\n",
    "kallDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TranscriptomeID</th>\n",
       "      <th>PDX_10Veh</th>\n",
       "      <th>PDX_12Veh</th>\n",
       "      <th>PDX_13P3_40</th>\n",
       "      <th>PDX_14P3_20</th>\n",
       "      <th>PDX_1P3_20</th>\n",
       "      <th>PDX_2P3_40</th>\n",
       "      <th>PDX_3P3_40</th>\n",
       "      <th>PDX_4P3_20</th>\n",
       "      <th>PDX_5P3_40</th>\n",
       "      <th>PDX_6P3_20</th>\n",
       "      <th>PDX_7P3_20</th>\n",
       "      <th>PDX_8Veh</th>\n",
       "      <th>PrimaryTumor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>3110.826328</td>\n",
       "      <td>1373.233983</td>\n",
       "      <td>4511.588795</td>\n",
       "      <td>3094.840621</td>\n",
       "      <td>3833.497701</td>\n",
       "      <td>4608.401041</td>\n",
       "      <td>3907.481643</td>\n",
       "      <td>4794.502657</td>\n",
       "      <td>5532.675937</td>\n",
       "      <td>3763.096813</td>\n",
       "      <td>3977.164745</td>\n",
       "      <td>4649.666301</td>\n",
       "      <td>4372.383189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000494424</td>\n",
       "      <td>20.942942</td>\n",
       "      <td>47.558741</td>\n",
       "      <td>63.957345</td>\n",
       "      <td>31.399837</td>\n",
       "      <td>11.134898</td>\n",
       "      <td>26.683315</td>\n",
       "      <td>19.393524</td>\n",
       "      <td>34.556585</td>\n",
       "      <td>48.489554</td>\n",
       "      <td>33.022034</td>\n",
       "      <td>51.166021</td>\n",
       "      <td>40.050277</td>\n",
       "      <td>39.141640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000496771</td>\n",
       "      <td>73.652784</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>114.277138</td>\n",
       "      <td>54.479622</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>105.946174</td>\n",
       "      <td>50.769851</td>\n",
       "      <td>99.604675</td>\n",
       "      <td>131.142126</td>\n",
       "      <td>81.204527</td>\n",
       "      <td>71.727135</td>\n",
       "      <td>90.073029</td>\n",
       "      <td>101.147416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000373031</td>\n",
       "      <td>9.199564</td>\n",
       "      <td>23.186203</td>\n",
       "      <td>23.203554</td>\n",
       "      <td>9.515067</td>\n",
       "      <td>5.582347</td>\n",
       "      <td>3.895577</td>\n",
       "      <td>2.241603</td>\n",
       "      <td>4.645070</td>\n",
       "      <td>7.746967</td>\n",
       "      <td>3.959939</td>\n",
       "      <td>6.100241</td>\n",
       "      <td>13.082332</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000485971</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>24.733969</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196349</th>\n",
       "      <td>ENST00000608041</td>\n",
       "      <td>9.270182</td>\n",
       "      <td>76.011547</td>\n",
       "      <td>5.335229</td>\n",
       "      <td>25.616037</td>\n",
       "      <td>24.324704</td>\n",
       "      <td>24.286108</td>\n",
       "      <td>22.277398</td>\n",
       "      <td>16.708099</td>\n",
       "      <td>9.101200</td>\n",
       "      <td>5.961555</td>\n",
       "      <td>5.372735</td>\n",
       "      <td>8.497456</td>\n",
       "      <td>46.326490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196350</th>\n",
       "      <td>ENST00000610193</td>\n",
       "      <td>18.372827</td>\n",
       "      <td>772.492540</td>\n",
       "      <td>2.327237</td>\n",
       "      <td>0.677210</td>\n",
       "      <td>36.920302</td>\n",
       "      <td>3.239746</td>\n",
       "      <td>20.949518</td>\n",
       "      <td>1.988206</td>\n",
       "      <td>1.289507</td>\n",
       "      <td>10.553138</td>\n",
       "      <td>12.946252</td>\n",
       "      <td>2.306854</td>\n",
       "      <td>142.329322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196351</th>\n",
       "      <td>ENST00000608591</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>206.432205</td>\n",
       "      <td>6.197407</td>\n",
       "      <td>2.713438</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.524030</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196352</th>\n",
       "      <td>ENST00000609365</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196353</th>\n",
       "      <td>ENST00000609225</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>196354 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        TranscriptomeID    PDX_10Veh    PDX_12Veh  PDX_13P3_40  PDX_14P3_20  \\\n",
       "0       ENST00000373020  3110.826328  1373.233983  4511.588795  3094.840621   \n",
       "1       ENST00000494424    20.942942    47.558741    63.957345    31.399837   \n",
       "2       ENST00000496771    73.652784     0.000000   114.277138    54.479622   \n",
       "3       ENST00000373031     9.199564    23.186203    23.203554     9.515067   \n",
       "4       ENST00000485971     0.000000    24.733969     0.000000     0.000000   \n",
       "...                 ...          ...          ...          ...          ...   \n",
       "196349  ENST00000608041     9.270182    76.011547     5.335229    25.616037   \n",
       "196350  ENST00000610193    18.372827   772.492540     2.327237     0.677210   \n",
       "196351  ENST00000608591     0.000000   206.432205     6.197407     2.713438   \n",
       "196352  ENST00000609365     0.000000     0.000000     0.000000     0.000000   \n",
       "196353  ENST00000609225     0.000000     0.000000     0.000000     0.000000   \n",
       "\n",
       "         PDX_1P3_20   PDX_2P3_40   PDX_3P3_40   PDX_4P3_20   PDX_5P3_40  \\\n",
       "0       3833.497701  4608.401041  3907.481643  4794.502657  5532.675937   \n",
       "1         11.134898    26.683315    19.393524    34.556585    48.489554   \n",
       "2          0.000000   105.946174    50.769851    99.604675   131.142126   \n",
       "3          5.582347     3.895577     2.241603     4.645070     7.746967   \n",
       "4          0.000000     0.000000     0.000000     0.000000     0.000000   \n",
       "...             ...          ...          ...          ...          ...   \n",
       "196349    24.324704    24.286108    22.277398    16.708099     9.101200   \n",
       "196350    36.920302     3.239746    20.949518     1.988206     1.289507   \n",
       "196351     0.000000     0.000000     0.000000     0.000000     0.000000   \n",
       "196352     0.000000     0.000000     0.000000     0.000000     0.000000   \n",
       "196353     0.000000     0.000000     0.000000     0.000000     0.000000   \n",
       "\n",
       "         PDX_6P3_20   PDX_7P3_20     PDX_8Veh  PrimaryTumor  \n",
       "0       3763.096813  3977.164745  4649.666301   4372.383189  \n",
       "1         33.022034    51.166021    40.050277     39.141640  \n",
       "2         81.204527    71.727135    90.073029    101.147416  \n",
       "3          3.959939     6.100241    13.082332      0.000000  \n",
       "4          0.000000     0.000000     0.000000      0.000000  \n",
       "...             ...          ...          ...           ...  \n",
       "196349     5.961555     5.372735     8.497456     46.326490  \n",
       "196350    10.553138    12.946252     2.306854    142.329322  \n",
       "196351     0.000000     1.524030     0.000000      0.000000  \n",
       "196352     0.000000     0.000000     0.000000      0.000000  \n",
       "196353     0.000000     0.000000     0.000000      0.000000  \n",
       "\n",
       "[196354 rows x 14 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rsemPath = \"/home/lima/Projects/simulation/TranscriptQuant.RealDataRSEM/cntMatrices_RSEM.csv\"\n",
    "rsemDF = pd.read_csv(rsemPath)\n",
    "rsemDF = rsemDF.rename(columns = {\"Unnamed: 0\": \"TranscriptomeID\"})\n",
    "rsemDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "kallDF['TranscriptomeID'] = kallDF['TranscriptomeID'].str.replace(\"\\.[0-9]\", \"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "kallTest = kallDF[['TranscriptomeID', 'PDX_10Veh_Kall']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "rsemTest = rsemDF[['TranscriptomeID', 'PDX_10Veh']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "mergedDF = rsemTest.join(kallTest.set_index('TranscriptomeID'), on = 'TranscriptomeID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TranscriptomeID</th>\n",
       "      <th>PDX_10Veh</th>\n",
       "      <th>PDX_10Veh_Kall</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>3110.826328</td>\n",
       "      <td>4331.283964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000494424</td>\n",
       "      <td>20.942942</td>\n",
       "      <td>37.930461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000496771</td>\n",
       "      <td>73.652784</td>\n",
       "      <td>101.621339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000373031</td>\n",
       "      <td>9.199564</td>\n",
       "      <td>12.946618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000485971</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196349</th>\n",
       "      <td>ENST00000608041</td>\n",
       "      <td>9.270182</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196350</th>\n",
       "      <td>ENST00000610193</td>\n",
       "      <td>18.372827</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196351</th>\n",
       "      <td>ENST00000608591</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196352</th>\n",
       "      <td>ENST00000609365</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196353</th>\n",
       "      <td>ENST00000609225</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>196354 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        TranscriptomeID    PDX_10Veh  PDX_10Veh_Kall\n",
       "0       ENST00000373020  3110.826328     4331.283964\n",
       "1       ENST00000494424    20.942942       37.930461\n",
       "2       ENST00000496771    73.652784      101.621339\n",
       "3       ENST00000373031     9.199564       12.946618\n",
       "4       ENST00000485971     0.000000        0.000000\n",
       "...                 ...          ...             ...\n",
       "196349  ENST00000608041     9.270182             NaN\n",
       "196350  ENST00000610193    18.372827             NaN\n",
       "196351  ENST00000608591     0.000000             NaN\n",
       "196352  ENST00000609365     0.000000             NaN\n",
       "196353  ENST00000609225     0.000000             NaN\n",
       "\n",
       "[196354 rows x 3 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mergedDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Int64Index([   443,    839,    840,    841,    842,    843,    844,    845,\n",
       "               846,    847,\n",
       "            ...\n",
       "            196343, 196344, 196346, 196347, 196348, 196349, 196350, 196351,\n",
       "            196352, 196353],\n",
       "           dtype='int64', length=32978)"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kallNaNIdx = mergedDF['PDX_10Veh_Kall'].index[mergedDF['PDX_10Veh_Kall'].apply(np.isnan)]\n",
    "kallNaNIdx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "kallNaNIdxList = kallNaNIdx.values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TranscriptomeID</th>\n",
       "      <th>PDX_10Veh</th>\n",
       "      <th>PDX_10Veh_Kall</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>443</th>\n",
       "      <td>ENST00000004531</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>839</th>\n",
       "      <td>ENST00000382189</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>840</th>\n",
       "      <td>ENST00000452401</td>\n",
       "      <td>60.963431</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>841</th>\n",
       "      <td>ENST00000585725</td>\n",
       "      <td>71.375940</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>842</th>\n",
       "      <td>ENST00000586332</td>\n",
       "      <td>2.524157</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196349</th>\n",
       "      <td>ENST00000608041</td>\n",
       "      <td>9.270182</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196350</th>\n",
       "      <td>ENST00000610193</td>\n",
       "      <td>18.372827</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196351</th>\n",
       "      <td>ENST00000608591</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196352</th>\n",
       "      <td>ENST00000609365</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196353</th>\n",
       "      <td>ENST00000609225</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32978 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        TranscriptomeID  PDX_10Veh  PDX_10Veh_Kall\n",
       "443     ENST00000004531   0.000000             NaN\n",
       "839     ENST00000382189   0.000000             NaN\n",
       "840     ENST00000452401  60.963431             NaN\n",
       "841     ENST00000585725  71.375940             NaN\n",
       "842     ENST00000586332   2.524157             NaN\n",
       "...                 ...        ...             ...\n",
       "196349  ENST00000608041   9.270182             NaN\n",
       "196350  ENST00000610193  18.372827             NaN\n",
       "196351  ENST00000608591   0.000000             NaN\n",
       "196352  ENST00000609365   0.000000             NaN\n",
       "196353  ENST00000609225   0.000000             NaN\n",
       "\n",
       "[32978 rows x 3 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mergedDF.loc[kallNaNIdxList]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "rsemNaNIdx = mergedDF['PDX_10Veh'].index[mergedDF['PDX_10Veh'].apply(np.isnan)]\n",
    "rsemNaNIdxList = rsemNaNIdx.values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TranscriptomeID</th>\n",
       "      <th>PDX_10Veh</th>\n",
       "      <th>PDX_10Veh_Kall</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [TranscriptomeID, PDX_10Veh, PDX_10Veh_Kall]\n",
       "Index: []"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mergedDF.loc[rsemNaNIdxList]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "rsemZeroIdx = mergedDF['PDX_10Veh'].index[mergedDF['PDX_10Veh'] == 0.0]\n",
    "rsemZeroIdxList = rsemZeroIdx.values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "kallZeroIdx = mergedDF['PDX_10Veh_Kall'].index[mergedDF['PDX_10Veh_Kall'] == 0.0]\n",
    "kallZeroIdxList = kallZeroIdx.values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "kallNaNIdxSet = set(kallNaNIdxList)\n",
    "rsemZeroIdxSet = set(rsemZeroIdxList)\n",
    "kallZeroIdxSet = set(kallZeroIdxList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "foundByRsemOnlyButZero = kallNaNIdxSet & rsemZeroIdxSet\n",
    "foundByRsemOnlyAndExpressed = kallNaNIdxSet - foundByRsemOnlyButZero\n",
    "foundByBothButZeroInBoth = rsemZeroIdxSet & kallZeroIdxSet\n",
    "foundByBothButZeroInRsem = rsemZeroIdxSet - foundByBothButZeroInBoth - foundByRsemOnlyButZero\n",
    "foundByBothButZeroInKall = kallZeroIdxSet - foundByBothButZeroInBoth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "foundByRsemOnlyButZero & foundByBothButZeroInRsem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [],
   "source": [
    "mergedDF['Label'] = 'FoundByBothAndExpreseedInBoth'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [],
   "source": [
    "mergedDF.loc[list(foundByRsemOnlyButZero), 'Label'] = \"FoundByRsemOnlyButZero\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [],
   "source": [
    "mergedDF.loc[list(foundByRsemOnlyAndExpressed), 'Label'] = \"FoundByRsemOnlyAndExpressed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [],
   "source": [
    "mergedDF.loc[list(foundByBothButZeroInBoth), 'Label'] = \"FoundByBothButZeroInBoth\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [],
   "source": [
    "mergedDF.loc[list(foundByBothButZeroInRsem), 'Label'] = \"FoundBytBothButZeroInRsem\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [],
   "source": [
    "mergedDF.loc[list(foundByBothButZeroInKall), 'Label'] = \"FoundByBothButZeroInKall\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [],
   "source": [
    "mergedDF['PDX_10Veh_Kall'] = mergedDF['PDX_10Veh_Kall'].fillna(0.0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = mergedDF.drop(columns = ['TranscriptomeID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[['PDX_10Veh', 'PDX_10Veh_Kall']] = np.log2(df[['PDX_10Veh', 'PDX_10Veh_Kall']] + 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0         FoundByBothAndExpreseedInBoth\n",
       "1         FoundByBothAndExpreseedInBoth\n",
       "2         FoundByBothAndExpreseedInBoth\n",
       "3         FoundByBothAndExpreseedInBoth\n",
       "4              FoundByBothButZeroInBoth\n",
       "                      ...              \n",
       "196349      FoundByRsemOnlyAndExpressed\n",
       "196350      FoundByRsemOnlyAndExpressed\n",
       "196351           FoundByRsemOnlyButZero\n",
       "196352           FoundByRsemOnlyButZero\n",
       "196353           FoundByRsemOnlyButZero\n",
       "Name: Label, Length: 196354, dtype: object"
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.Label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(style = \"whitegrid\", palette = \"muted\", color_codes = True)\n",
    "figureOne, axes = plt.subplots(2, 2, constrained_layout = True, figsize = (20, 20))\n",
    "# put the violin plot to axes[0, 0]\n",
    "axes[0, 0].clear()\n",
    "sns.violinplot(data = df, linewidth = 0.8, palette = \"husl\", ax = axes[0, 0])\n",
    "    \n",
    "# put the boxplot to axes[1, 1]\n",
    "axes[1, 1].clear()\n",
    "sns.boxplot(data = df, linewidth = 0.8, palette = \"husl\", ax = axes[1, 1])\n",
    "    \n",
    "# put the distribution plot to axes[0, 1]\n",
    "axes[0, 1].clear()\n",
    "sns.distplot(df.iloc[ : , 0], kde = True, color = \"b\", label = 'RSEM', ax = axes[0, 1])\n",
    "sns.distplot(df.iloc[ : , 1], kde = True, color = \"r\", label = 'Kallisto', ax = axes[0, 1])\n",
    "axes[0, 1].legend(loc = 'best', prop = {'size': 10})\n",
    "axes[0, 1].set_xlabel(\"Distribution plot for counts found by rsem and counts found by kall\")\n",
    "\n",
    "sns.scatterplot(x = 'PDX_10Veh', y = 'PDX_10Veh_Kall', hue = 'Label', marker = '+', data = df, ax = axes[1, 0], palette = 'muted', s = 2.0)\n",
    "#at = AnchoredText(\"{inBoth} are found in both real and simulated data\\n{neither} expressed in neither real nor simulated data\\n{realOnly} are found in real data only\\n{simOnly} are found in simulated data only\\nCut off threshould: {cut}\\nCorr: {corr}\".format(\n",
    "#    neither = neither, realOnly = realOnly, simOnly = simOnly, inBoth = inBoth, corr = corr, total = len(df) - realOnly - simOnly + len(inNeither), cut = cutOff), \n",
    "#                  prop = dict(size = 8), frameon = True, loc = 'upper left')\n",
    "#at.patch.set_boxstyle(\"round,pad=0.,rounding_size=0.2\")\n",
    "   \n",
    "#axes[1, 0].add_artist(at)\n",
    "axes[1, 0].set_xlabel(\"RSEM\")\n",
    "axes[1, 0].set_ylabel(\"Kallisto\")\n",
    "axes[1, 0].legend(loc = 'lower right', fancybox = True, shadow = False, scatterpoints = 200, prop = {'size': 8})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "CONENV",
   "language": "python",
   "name": "conenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
