{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scipy.stats\n",
    "import matplotlib.pylab as pylab\n",
    "import ipywidgets as widgets\n",
    "from os import path\n",
    "from sklearn.decomposition import PCA\n",
    "from matplotlib.offsetbox import AnchoredText\n",
    "from IPython.display import clear_output\n",
    "from IPython.display import display\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "class DataLoader:\n",
    "    def __init__(self):\n",
    "        self.__kall = None\n",
    "        self.__rsem = {}\n",
    "        \n",
    "    def __loadKall(self):\n",
    "        kallDF = pd.read_csv('/home/lima/Project/simulation/Comparision/Kall/Real/KallistoReadCounts.csv')\n",
    "        kallDF = kallDF.rename(columns = {\"Unnamed: 0\": \"GeneID\"})\n",
    "        kallDF.set_index(\"GeneID\", inplace = True)\n",
    "        self.__kall = kallDF\n",
    "    \n",
    "    def __loadRSEM(self):\n",
    "        cutOff = [0, 2.5, 5, 7.5, 10]\n",
    "        # key of the tempDict is the cut off threshould, value is the cnt matrix accordingly\n",
    "        tempDict = {}\n",
    "        dictList = []\n",
    "        for cut in cutOff:            \n",
    "            fullPath = \"/home/lima/Project/simulation/Comparision/RSEM/Real/CountsCutOffValue{cut}.csv\".format(cut = cut)\n",
    "            dataFrame = pd.read_csv(fullPath)\n",
    "            dataFrame = dataFrame.rename(columns = {\"Unnamed: 0\": \"GeneID\"})\n",
    "            dataFrame.set_index(\"GeneID\", inplace = True)\n",
    "            \n",
    "            tempDict[cut] = dataFrame\n",
    "        self.__rsem = tempDict\n",
    "    \n",
    "    def getKall(self):\n",
    "        self.__loadKall()\n",
    "        return self.__kall\n",
    "    \n",
    "    def getRSEM(self):\n",
    "        self.__loadRSEM()\n",
    "        return self.__rsem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ConcateData:\n",
    "    def __init__(self, groupList, rsem, kallisto):\n",
    "        # input:\n",
    "        # @groupList: a list contains the names of samples. there are four groups of samples, high, low, veh, pt\n",
    "        # @rsem: rsem counts\n",
    "        # @kallisto: kallisto counts\n",
    "        # both counts are in gene level\n",
    "        self.__list = groupList\n",
    "        self.__rsem = rsem\n",
    "        self.__kallisto = kallisto\n",
    "        self.__result = []\n",
    "        \n",
    "    def __concat(self):        \n",
    "        \n",
    "        for sample in self.__list:\n",
    "            targetRSEM = self.__rsem[[sample]]\n",
    "            targetKall = self.__kallisto[[sample]]\n",
    "            # print(targetRSEM)\n",
    "            # print(targetKall)\n",
    "            rsemLen = len(targetRSEM)\n",
    "            kallLen = len(targetKall)\n",
    "        \n",
    "            print(\"{num} genes have been detected in {sample} by RSEM\".format(num = rsemLen, sample = sample))\n",
    "            print(\"{num} genes have been detected in {sample} by kallisto\".format(num = kallLen, sample = sample))\n",
    "            \n",
    "            # convert the index, gene ids, to the first column for merge two dataframes\n",
    "            targetRSEM.reset_index(inplace = True)\n",
    "            targetKall.reset_index(inplace = True)\n",
    "            # print(targetKall)\n",
    "            # print(targetRSEM)\n",
    "            df = pd.merge(targetRSEM, targetKall, how = 'outer', on = 'GeneID', suffixes = ('_RSEM', '_Kallisto'))\n",
    "            print(\"{num} genes in both dataframes\".format(num = len(df)))\n",
    "            # take out the GeneIDs in each data frame and put them into two sets\n",
    "            rsemIdx = set(targetRSEM['GeneID'])\n",
    "            kallIdx = set(targetKall['GeneID'])\n",
    "            \n",
    "            inBoth = rsemIdx & kallIdx\n",
    "            rsemOnly = rsemIdx - inBoth\n",
    "            kallOnly = kallIdx - inBoth\n",
    "            \n",
    "            inBoth = list(inBoth)\n",
    "            rsemOnly = list(rsemOnly)\n",
    "            kallOnly = list(kallOnly)\n",
    "            \n",
    "            bothDF = pd.DataFrame(list(zip(inBoth, [\"DetectedByBoth\"] * len(inBoth))), columns = [\"GeneID\", \"Label\"])\n",
    "            rsemOnlyDF = pd.DataFrame(list(zip(rsemOnly, [\"OnlyDetectedByRSEM\"] * len(rsemOnly))), columns = [\"GeneID\", \"Label\"])\n",
    "            kallOnlyDF = pd.DataFrame(list(zip(kallOnly, [\"OnlyDetectedByKallisto\"] * len(kallOnly))), columns = [\"GeneID\", \"Label\"])\n",
    "            labelDF = pd.concat([bothDF, rsemOnlyDF, kallOnlyDF])\n",
    "            print(\"{num} labels have been attached\".format(num = len(labelDF)))\n",
    "            # fill nans in df with 0\n",
    "            df.fillna(0, inplace = True)\n",
    "            df = pd.merge(df, labelDF, how = 'inner', on = 'GeneID')\n",
    "            #print(df)\n",
    "            self.__result.append(df)\n",
    "            \n",
    "    def setData(self):\n",
    "        self.__concat()\n",
    "        \n",
    "    def getData(self):\n",
    "        return self.__result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Plotter:\n",
    "    def __init__(self, pairedData):\n",
    "        self.__data = pairedData\n",
    "        \n",
    "    def plotPlotBlock(self, df):\n",
    "        \n",
    "        params = {'legend.fontsize': 8, \n",
    "                    'font.family': 'serif',\n",
    "                    'font.weight': 'medium',\n",
    "                    'font.variant': 'normal',\n",
    "                    'figure.figsize': (10, 10), \n",
    "                    'figure.edgecolor': '#04253a', \n",
    "                    'figure.titlesize': 8, \n",
    "                    'axes.labelsize': 6, \n",
    "                    'axes.titlesize': 6, \n",
    "                    'xtick.labelsize': 6, \n",
    "                    'ytick.labelsize': 6}\n",
    "        pylab.rcParams.update(params)\n",
    "        figureOne, axes = plt.subplots(2, 2, constrained_layout = True)\n",
    "        # self.buildPlotBlock(df, axes, cutOff, names)\n",
    "        self.buildPlotBlock(df, axes)\n",
    "        # plt.suptitle('Statistical plots for comparing {realData} & {simulatedData} with reads shorter than {cut} being removed'.format(realData = names[0], simulatedData = names[1], cut = cutOff))\n",
    "        # title = 'Real data and synthetic data of {realData} with reads shorter than {cut} being removed.png'.format(realData = names[0], cut = cutOff)\n",
    "        plt.savefig(title, dpi = 300, bbox_inches = 'tight')\n",
    "        plt.plot()\n",
    "        \n",
    "    def buildPlotBlock(self, df, axes):\n",
    "        \"\"\"\n",
    "        buildPlotBlock is a helper function helping putting the desired plots, such as the scatter plot, box plot,\n",
    "        violin plot and the distribution plot into a grided pyplot\n",
    "    \n",
    "        Input:\n",
    "        @df: the dataframe which provides the data\n",
    "        @axes: a 2x2 subplot matrix \n",
    "        \"\"\"\n",
    "        # df = np.log2(df[names] + 1)\n",
    "        sns.set(style = \"whitegrid\", palette = \"muted\", color_codes = True)\n",
    "        # put the violin plot to axes[0, 0]\n",
    "        axes[0, 0].clear()\n",
    "        sns.violinplot(data = df, linewidth = 0.2, palette = \"husl\", ax = axes[0, 0])\n",
    "    \n",
    "        # put the boxplot to axes[1, 1]\n",
    "        axes[1, 1].clear()\n",
    "        sns.boxplot(data = df, linewidth = 0.2, palette = \"husl\", ax = axes[1, 1])\n",
    "    \n",
    "        # put the distribution plot to axes[0, 1]\n",
    "        axes[0, 1].clear()\n",
    "        sns.distplot(df.iloc[ : , 0], kde = True, color = \"b\", label = 'Real', ax = axes[0, 1])\n",
    "        sns.distplot(df.iloc[ : , 1], kde = True, color = \"r\", label = 'Simulated', ax = axes[0, 1])\n",
    "        axes[0, 1].legend(loc = 'best', prop = {'size': 8})\n",
    "        axes[0, 1].set_xlabel(\"Distribution plot for real data and simulated data\")\n",
    "    \n",
    "        # put the scatter plot to axes [1, 0]\n",
    "        axes[1, 0].clear()\n",
    "        self.scatterPlot(df, axes[1, 0], cutOff, names)\n",
    "                         \n",
    "    def scatterPlot(self, df, ax):\n",
    "        \"\"\"\n",
    "        the scatterPlot function plots a pairwise scatter plot for the simulated data and the real data for the same \n",
    "        sample, with the x-coordinates being the counts number of the real data and the y-coordinates being the counts\n",
    "        number of the simulated data\n",
    "    \n",
    "        input data:\n",
    "        @df: a pandas dataframe contains the counts for the real and the simulated data\n",
    "        @ax: an axes object\n",
    "        @names: a list contains the sample names\n",
    "        \"\"\"\n",
    "    \n",
    "        # notInrealData, notInSimulatedData, inNeither are lists stores the row indecies for data as theirs names\n",
    "        # notInRealData = df.index[df.iloc[ : , 0] == 0].tolist()\n",
    "        # notInSimulatedData = df.index[df.iloc[ : , 1] == 0].tolist()\n",
    "        # inNeither = list(set(notInRealData) & set(notInSimulatedData))\n",
    "    \n",
    "        # inBoth, simOnly, realOnly, neither are integers for \n",
    "        # genes found in both samples, one of the samples or none of them\n",
    "        # neither = len(inNeither)\n",
    "        # simOnly = len(notInRealData) - neither\n",
    "        # realOnly = len(notInSimulatedData) - neither\n",
    "        # inBoth = len(df) - simOnly - realOnly + neither\n",
    "    \n",
    "        # category the genes into 4 classes\n",
    "        # df[\"ExpressionStatus\"] = [\"BothSamples\"] * len(df)\n",
    "        # df.loc[notInRealData, \"ExpressionStatus\"] = \"SimulatedOnly\"\n",
    "        # df.loc[notInSimulatedData, \"ExpressionStatus\"] = \"RealOnly\"\n",
    "        # df.loc[inNeither, \"ExpressionStatus\"] = \"NeithSample\"\n",
    "        # print(df)\n",
    "        # set up the plotting\n",
    "        sns.set(style = \"whitegrid\", palette = \"muted\", color_codes = True)\n",
    "        # sns.set(style = \"dark\", palette = \"inferno\", color_codes = True)\n",
    "        # calculate the correlation between the two sets of data.\n",
    "        # print(df.iloc[ : , 0])\n",
    "        names = list(df.columns)\n",
    "        col_one = df.iloc[ : , 1]\n",
    "        col_two = df.iloc[ : , 2]\n",
    "        # corr = scipy.stats.pearsonr(col_one, col_two)[0]\n",
    "        \n",
    "        sns.scatterplot(x = names[1], y = names[2], hue = 'Lables', marker = '+', style = 'ExpressionStatus', data = df, ax = ax, palette = 'inferno', s = 2.0)\n",
    "        at = AnchoredText(\"{inBoth} are found in both real and simulated data\\n{neither} expressed in neither real nor simulated data\\n{realOnly} are found in real data only\\n{simOnly} are found in simulated data only\\nCut off threshould: {cut}\\nCorr: {corr}\".format(\n",
    "            neither = neither, realOnly = realOnly, simOnly = simOnly, inBoth = inBoth, corr = corr, total = len(df) - realOnly - simOnly + len(inNeither), cut = cutOff), \n",
    "                          prop = dict(size = 8), frameon = True, loc = 'upper left')\n",
    "        at.patch.set_boxstyle(\"round,pad=0.,rounding_size=0.2\")\n",
    "   \n",
    "        ax.add_artist(at)\n",
    "        ax.set_xlabel(\"Real Data\")\n",
    "        ax.set_ylabel(\"Simulated Data\")\n",
    "        ax.legend(loc = 'lower right', fancybox = True, shadow = False, scatterpoints = 200, prop = {'size': 8})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "loader = DataLoader()\n",
    "kall = loader.getKall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "rsem = loader.getRSEM()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = rsem[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PDX_10Veh</th>\n",
       "      <th>PDX_12Veh</th>\n",
       "      <th>PDX_13P3_40</th>\n",
       "      <th>PDX_14P3_20</th>\n",
       "      <th>PDX_1P3_20</th>\n",
       "      <th>PDX_2P3_40</th>\n",
       "      <th>PDX_3P3_40</th>\n",
       "      <th>PDX_4P3_20</th>\n",
       "      <th>PDX_5P3_40</th>\n",
       "      <th>PDX_6P3_20</th>\n",
       "      <th>PDX_7P3_20</th>\n",
       "      <th>PDX_8Veh</th>\n",
       "      <th>PrimaryTumor</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GeneID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>11.986054</td>\n",
       "      <td>10.654686</td>\n",
       "      <td>12.632209</td>\n",
       "      <td>11.984520</td>\n",
       "      <td>12.154459</td>\n",
       "      <td>12.590210</td>\n",
       "      <td>12.347698</td>\n",
       "      <td>12.682657</td>\n",
       "      <td>12.871223</td>\n",
       "      <td>12.345260</td>\n",
       "      <td>12.411518</td>\n",
       "      <td>12.649463</td>\n",
       "      <td>12.064354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <td>6.239048</td>\n",
       "      <td>8.567898</td>\n",
       "      <td>6.607359</td>\n",
       "      <td>6.334044</td>\n",
       "      <td>6.146942</td>\n",
       "      <td>6.083594</td>\n",
       "      <td>5.886587</td>\n",
       "      <td>6.250034</td>\n",
       "      <td>6.234144</td>\n",
       "      <td>6.058683</td>\n",
       "      <td>6.101227</td>\n",
       "      <td>6.355960</td>\n",
       "      <td>6.591030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>10.992695</td>\n",
       "      <td>8.771258</td>\n",
       "      <td>11.214416</td>\n",
       "      <td>10.664481</td>\n",
       "      <td>10.790173</td>\n",
       "      <td>11.174510</td>\n",
       "      <td>11.051685</td>\n",
       "      <td>11.486012</td>\n",
       "      <td>11.258944</td>\n",
       "      <td>11.095102</td>\n",
       "      <td>10.942841</td>\n",
       "      <td>11.140087</td>\n",
       "      <td>9.527697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>9.080455</td>\n",
       "      <td>10.036320</td>\n",
       "      <td>8.627819</td>\n",
       "      <td>8.391202</td>\n",
       "      <td>8.666030</td>\n",
       "      <td>8.619614</td>\n",
       "      <td>8.353345</td>\n",
       "      <td>8.312060</td>\n",
       "      <td>8.457892</td>\n",
       "      <td>8.509969</td>\n",
       "      <td>8.979662</td>\n",
       "      <td>8.681332</td>\n",
       "      <td>10.249468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>9.475529</td>\n",
       "      <td>11.102118</td>\n",
       "      <td>8.738020</td>\n",
       "      <td>8.628993</td>\n",
       "      <td>9.447304</td>\n",
       "      <td>8.912947</td>\n",
       "      <td>8.809385</td>\n",
       "      <td>8.929883</td>\n",
       "      <td>8.723530</td>\n",
       "      <td>9.074169</td>\n",
       "      <td>9.216291</td>\n",
       "      <td>8.634262</td>\n",
       "      <td>10.227561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000273475</th>\n",
       "      <td>5.551952</td>\n",
       "      <td>6.756068</td>\n",
       "      <td>5.551952</td>\n",
       "      <td>5.551952</td>\n",
       "      <td>5.551952</td>\n",
       "      <td>5.551952</td>\n",
       "      <td>5.551952</td>\n",
       "      <td>5.551952</td>\n",
       "      <td>5.551952</td>\n",
       "      <td>5.814314</td>\n",
       "      <td>5.551952</td>\n",
       "      <td>5.551952</td>\n",
       "      <td>6.375848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000273479</th>\n",
       "      <td>5.551952</td>\n",
       "      <td>6.546476</td>\n",
       "      <td>5.551952</td>\n",
       "      <td>5.551952</td>\n",
       "      <td>5.551952</td>\n",
       "      <td>5.551952</td>\n",
       "      <td>5.551952</td>\n",
       "      <td>5.740078</td>\n",
       "      <td>5.551952</td>\n",
       "      <td>5.737973</td>\n",
       "      <td>5.551952</td>\n",
       "      <td>5.551952</td>\n",
       "      <td>5.871356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000273482</th>\n",
       "      <td>9.862638</td>\n",
       "      <td>8.866822</td>\n",
       "      <td>9.982666</td>\n",
       "      <td>10.061316</td>\n",
       "      <td>9.863262</td>\n",
       "      <td>9.741165</td>\n",
       "      <td>9.859578</td>\n",
       "      <td>9.734819</td>\n",
       "      <td>9.939446</td>\n",
       "      <td>9.771534</td>\n",
       "      <td>9.946812</td>\n",
       "      <td>9.856527</td>\n",
       "      <td>9.504428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000273484</th>\n",
       "      <td>5.551952</td>\n",
       "      <td>6.990805</td>\n",
       "      <td>5.551952</td>\n",
       "      <td>6.417861</td>\n",
       "      <td>5.883522</td>\n",
       "      <td>5.874935</td>\n",
       "      <td>5.551952</td>\n",
       "      <td>5.551952</td>\n",
       "      <td>5.814254</td>\n",
       "      <td>5.551952</td>\n",
       "      <td>5.754059</td>\n",
       "      <td>6.089207</td>\n",
       "      <td>6.913152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000273490</th>\n",
       "      <td>8.284557</td>\n",
       "      <td>5.551952</td>\n",
       "      <td>8.538644</td>\n",
       "      <td>8.483419</td>\n",
       "      <td>8.147020</td>\n",
       "      <td>7.853319</td>\n",
       "      <td>8.296541</td>\n",
       "      <td>8.154758</td>\n",
       "      <td>8.074652</td>\n",
       "      <td>7.913044</td>\n",
       "      <td>7.477447</td>\n",
       "      <td>8.258490</td>\n",
       "      <td>7.192861</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>39293 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 PDX_10Veh  PDX_12Veh  PDX_13P3_40  PDX_14P3_20  PDX_1P3_20  \\\n",
       "GeneID                                                                        \n",
       "ENSG00000000003  11.986054  10.654686    12.632209    11.984520   12.154459   \n",
       "ENSG00000000005   6.239048   8.567898     6.607359     6.334044    6.146942   \n",
       "ENSG00000000419  10.992695   8.771258    11.214416    10.664481   10.790173   \n",
       "ENSG00000000457   9.080455  10.036320     8.627819     8.391202    8.666030   \n",
       "ENSG00000000460   9.475529  11.102118     8.738020     8.628993    9.447304   \n",
       "...                    ...        ...          ...          ...         ...   \n",
       "ENSG00000273475   5.551952   6.756068     5.551952     5.551952    5.551952   \n",
       "ENSG00000273479   5.551952   6.546476     5.551952     5.551952    5.551952   \n",
       "ENSG00000273482   9.862638   8.866822     9.982666    10.061316    9.863262   \n",
       "ENSG00000273484   5.551952   6.990805     5.551952     6.417861    5.883522   \n",
       "ENSG00000273490   8.284557   5.551952     8.538644     8.483419    8.147020   \n",
       "\n",
       "                 PDX_2P3_40  PDX_3P3_40  PDX_4P3_20  PDX_5P3_40  PDX_6P3_20  \\\n",
       "GeneID                                                                        \n",
       "ENSG00000000003   12.590210   12.347698   12.682657   12.871223   12.345260   \n",
       "ENSG00000000005    6.083594    5.886587    6.250034    6.234144    6.058683   \n",
       "ENSG00000000419   11.174510   11.051685   11.486012   11.258944   11.095102   \n",
       "ENSG00000000457    8.619614    8.353345    8.312060    8.457892    8.509969   \n",
       "ENSG00000000460    8.912947    8.809385    8.929883    8.723530    9.074169   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "ENSG00000273475    5.551952    5.551952    5.551952    5.551952    5.814314   \n",
       "ENSG00000273479    5.551952    5.551952    5.740078    5.551952    5.737973   \n",
       "ENSG00000273482    9.741165    9.859578    9.734819    9.939446    9.771534   \n",
       "ENSG00000273484    5.874935    5.551952    5.551952    5.814254    5.551952   \n",
       "ENSG00000273490    7.853319    8.296541    8.154758    8.074652    7.913044   \n",
       "\n",
       "                 PDX_7P3_20   PDX_8Veh  PrimaryTumor  \n",
       "GeneID                                                \n",
       "ENSG00000000003   12.411518  12.649463     12.064354  \n",
       "ENSG00000000005    6.101227   6.355960      6.591030  \n",
       "ENSG00000000419   10.942841  11.140087      9.527697  \n",
       "ENSG00000000457    8.979662   8.681332     10.249468  \n",
       "ENSG00000000460    9.216291   8.634262     10.227561  \n",
       "...                     ...        ...           ...  \n",
       "ENSG00000273475    5.551952   5.551952      6.375848  \n",
       "ENSG00000273479    5.551952   5.551952      5.871356  \n",
       "ENSG00000273482    9.946812   9.856527      9.504428  \n",
       "ENSG00000273484    5.754059   6.089207      6.913152  \n",
       "ENSG00000273490    7.477447   8.258490      7.192861  \n",
       "\n",
       "[39293 rows x 13 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "57773 genes have been detected in PDX_12Veh by RSEM\n",
      "39293 genes have been detected in PDX_12Veh by kallisto\n",
      "62158 genes in both dataframes\n",
      "62158 labels have been attached\n",
      "                GeneID  PDX_12Veh_RSEM  PDX_12Veh_Kallisto  \\\n",
      "0      ENSG00000000003            62.0           10.654686   \n",
      "1      ENSG00000000005             2.0            8.567898   \n",
      "2      ENSG00000000419            27.0            8.771258   \n",
      "3      ENSG00000000457            22.0           10.036320   \n",
      "4      ENSG00000000460            13.0           11.102118   \n",
      "...                ...             ...                 ...   \n",
      "62153  ENSG00000273470             0.0            6.995122   \n",
      "62154  ENSG00000273475             0.0            6.756068   \n",
      "62155  ENSG00000273479             0.0            6.546476   \n",
      "62156  ENSG00000273482             0.0            8.866822   \n",
      "62157  ENSG00000273490             0.0            5.551952   \n",
      "\n",
      "                        Label  \n",
      "0              DetectedByBoth  \n",
      "1              DetectedByBoth  \n",
      "2              DetectedByBoth  \n",
      "3              DetectedByBoth  \n",
      "4              DetectedByBoth  \n",
      "...                       ...  \n",
      "62153  OnlyDetectedByKallisto  \n",
      "62154  OnlyDetectedByKallisto  \n",
      "62155  OnlyDetectedByKallisto  \n",
      "62156  OnlyDetectedByKallisto  \n",
      "62157  OnlyDetectedByKallisto  \n",
      "\n",
      "[62158 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "concat = ConcateData(['PDX_12Veh'], test, kall)\n",
    "concat.setData()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "CONENV",
   "language": "python",
   "name": "conenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
